keps/sig-network/3015-prefer-local-traffic-policy/README.md

```diff
@@ -0,0 +1,548 @@
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `Local` policy for `service.spec.externalTrafficPolicy` and
+`service.spec.internalTrafficPolicy` allows configuring services so
+that they will always deliver to local endpoints, avoiding an
+unnecessary hop to another node when the request could just as easily
+be processed locally.
+
+However, they both fail entirely if there are no endpoints for the
+service on the node. This tradeoff is unacceptable for some use cases.
+
+This KEP adds a `PreferLocal` policy to both `externalTrafficPolicy`
+and `internalTrafficPolicy`, which minimizes hops when possible, but
+falls back when not.
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections from outside the
+  cluster will be delivered to a local endpoint when possible, and a
+  remote endpoint if not.
+
+- Allow configuring a service so that connections from inside the
+  cluster will be delivered to a local endpoint when possible, and a
+  remote endpoint if not.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### User Stories
+
+#### `PreferLocal` External Traffic Policy as a More-Reliable "Local"
+
+As an application developer, I want incoming service traffic from
+outside the cluster to be delivered to my pods as efficiently as
+possible (as with `Local` traffic policy), but without interruptions
+when endpoints are restarted or moved between nodes.
+
+The [Terminating Endpoints] feature was intended to make
+`externalTrafficPolicy: Local` services more robust when endpoints
+change, but it doesn't close the race conditions entirely: in
+particular, it only works if the time it takes the endpoint pod to
+terminate is longer than the amount of time between load balancer
+health checks.
+
+[Terminating Endpoints]: ../1669-graceful-termination-local-external-traffic-policy/
+
+#### `PreferLocal` Internal Traffic Policy as a More-Reliable `Local`
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should just go to a remote pod instead so it keeps working.
+
+(This is a modified version of a user story from the [Internal Traffic
```

Comment:
The original KEP actually proposed PreferLocal as one of the policy types, but we later removed it because we thought topology-aware routing could provide the same behavior with less complexity. Curious what changed for us to recondier PreferLocal again?  

(sorry if we already discussed this somewhere else, catching up on some context I missed)

Reply:
Sorry, spoke too soon! I see the section "### Interaction with Topology" now, will leave commens there